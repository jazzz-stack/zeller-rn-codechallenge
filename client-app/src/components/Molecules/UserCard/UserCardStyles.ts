import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  card: {
    backgroundColor: '#FFFFFF',
    paddingVertical: 16,
    paddingLeft: 0,
    marginBottom: 8,
    marginHorizontal: 16,
    borderBottomWidth: 1,
    borderColor: '#E0E0E0',
  },
  userInfo: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  userDetails: {
    flex: 1,
  },
  userDetailsStyle: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  userNameContainer: {
    height: 50,
    width: 50,
    borderRadius: 14,
    backgroundColor: '#e8f2fb',
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 16,
  },
  userInitials: {
    fontSize: 16,
    fontWeight: '700',
    color: '#0171ce',
    marginBottom: 4,
  },
  userName: {
    fontSize: 14,
    fontWeight: 'bold',
    color: '#2C3E50',
    marginBottom: 4,
  },
  roleText: {
    color: '#7F8C8D',
    fontSize: 14,
    fontWeight: '600',
  },
  actions: {
    flexDirection: 'row',
    gap: 10,
  },
  actionButton: {
    width: 30,
    height: 30,
    borderRadius: 18,
    justifyContent: 'center',
    alignItems: 'center',
  },
  editButton: {
    backgroundColor: '#e8f2fb',
    borderColor: '#0171ce',
    borderWidth: 1,
  },
  deleteButton: {
    backgroundColor: '#FDEDED',
    borderColor: '#E74C3C',
    borderWidth: 1,
  },
});
