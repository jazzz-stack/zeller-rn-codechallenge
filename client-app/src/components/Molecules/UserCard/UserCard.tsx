import React from 'react';
import {View, TouchableOpacity} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {styles} from './UserCardStyles';
import {UserCardProps} from '../schema';
import {Text} from '../../Atoms';

const UserCard: React.FC<UserCardProps> = ({user, onEdit, onDelete}) => {
  const getInitials = (name: string) => {
    const words = name.split(' ');
    const firstInitial = words[0]?.[0] || '';
    const lastInitial = words[1]?.[0] || '';
    return (firstInitial + lastInitial).toUpperCase();
  };

  return (
    <View style={styles.card}>
      <View style={styles.userInfo}>
        <View style={styles.userDetailsStyle}>
          <View style={styles.userNameContainer}>
            <Text style={styles.userInitials}>{getInitials(user.name)}</Text>
          </View>
          <View style={styles.userDetails}>
            <Text style={styles.userName}>{user.name}</Text>
            <Text style={styles.roleText}>{user.role}</Text>
          </View>
        </View>
        <View style={styles.actions}>
          <TouchableOpacity
            testID="usercard-edit"
            style={[styles.actionButton, styles.editButton]}
            onPress={() => onEdit(user)}>
            <Icon name="edit" size={18} color="#0171ce" />
          </TouchableOpacity>
          <TouchableOpacity
            testID="usercard-delete"
            style={[styles.actionButton, styles.deleteButton]}
            onPress={() => onDelete(user)}>
            <Icon name="delete" size={18} color="#E74C3C" />
          </TouchableOpacity>
        </View>
      </View>
    </View>
  );
};

export default UserCard;
