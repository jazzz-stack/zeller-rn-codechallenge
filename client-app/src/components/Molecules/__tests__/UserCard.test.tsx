import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react-native';
import {TouchableOpacity} from 'react-native';
import UserCard from '../UserCard/UserCard';

// Mock react-native-vector-icons
jest.mock('react-native-vector-icons/MaterialIcons', () => 'Icon');

// Mock the Text component
jest.mock('../../Atoms', () => ({
  Text: ({children, style, ...props}: any) => {
    const {Text} = require('react-native');
    return <Text style={style} {...props}>{children}</Text>;
  },
}));

describe('UserCard', () => {
  const mockUser = {
    id: '1',
    name: 'John Doe',
    email: 'john@example.com',
    role: 'Admin' as const,
  };

  const defaultProps = {
    user: mockUser,
    onEdit: jest.fn(),
    onDelete: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders user information', () => {
    render(<UserCard {...defaultProps} />);
    
    expect(screen.getByText('John Doe')).toBeTruthy();
    expect(screen.getByText('Admin')).toBeTruthy();
  });

  it('displays user initials correctly', () => {
    render(<UserCard {...defaultProps} />);
    
    expect(screen.getByText('JD')).toBeTruthy();
  });

  it('handles single name correctly', () => {
    const singleNameUser = {
      ...mockUser,
      name: 'John',
    };
    
    render(<UserCard {...defaultProps} user={singleNameUser} />);
    
    expect(screen.getByText('J')).toBeTruthy();
  });

  it('calls onEdit when edit button is pressed', () => {
    const {UNSAFE_root} = render(<UserCard {...defaultProps} />);
    
    const touchableElements = UNSAFE_root.findAllByType(TouchableOpacity);
    const editButton = touchableElements[0]; // First TouchableOpacity is edit button
    
    fireEvent.press(editButton);
    expect(defaultProps.onEdit).toHaveBeenCalledWith(mockUser);
  });

  it('calls onDelete when delete button is pressed', () => {
    const {UNSAFE_root} = render(<UserCard {...defaultProps} />);
    
    const touchableElements = UNSAFE_root.findAllByType(TouchableOpacity);
    const deleteButton = touchableElements[1]; // Second TouchableOpacity is delete button
    
    fireEvent.press(deleteButton);
    expect(defaultProps.onDelete).toHaveBeenCalledWith(mockUser);
  });

  // Snapshot Tests
  it('matches snapshot with admin user', () => {
    const component = render(<UserCard {...defaultProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with manager user', () => {
    const managerUser = {
      ...mockUser,
      role: 'Manager' as const,
    };
    const component = render(<UserCard {...defaultProps} user={managerUser} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('getInitials edge cases', () => {
    it('handles empty name', () => {
      const emptyNameUser = {
        ...mockUser,
        name: '',
      };
      
      render(<UserCard {...defaultProps} user={emptyNameUser} />);
      
      // Component should render without crashing for empty name
      expect(screen.getByText('Admin')).toBeTruthy(); // Role should still be visible
    });

    it('handles name with only spaces', () => {
      const spacesOnlyUser = {
        ...mockUser,
        name: '   ',
      };
      
      render(<UserCard {...defaultProps} user={spacesOnlyUser} />);
      
      // Component should render without crashing for spaces only
      expect(screen.getByText('Admin')).toBeTruthy(); // Role should still be visible
    });

    it('handles single character name', () => {
      const singleCharUser = {
        ...mockUser,
        name: 'A',
      };
      
      render(<UserCard {...defaultProps} user={singleCharUser} />);
      
      // Should show just the single character as initial
      expect(screen.getAllByText('A').length).toBeGreaterThan(0);
    });

    it('handles names with multiple spaces (treats as single space)', () => {
      const multiSpaceUser = {
        ...mockUser,
        name: 'John    Doe',
      };
      
      render(<UserCard {...defaultProps} user={multiSpaceUser} />);
      
      // Multiple spaces create empty array elements, so only first character is taken
      expect(screen.getByText('J')).toBeTruthy();
    });

    it('handles names with leading and trailing spaces', () => {
      const trimSpaceUser = {
        ...mockUser,
        name: '  John Doe  ',
      };
      
      render(<UserCard {...defaultProps} user={trimSpaceUser} />);
      
      // Should render without crashing and show role
      expect(screen.getByText('Admin')).toBeTruthy();
    });

    it('handles three or more names (takes first and second)', () => {
      const threeNamesUser = {
        ...mockUser,
        name: 'John Michael Doe',
      };
      
      render(<UserCard {...defaultProps} user={threeNamesUser} />);
      
      expect(screen.getByText('JM')).toBeTruthy();
    });
  });
});
