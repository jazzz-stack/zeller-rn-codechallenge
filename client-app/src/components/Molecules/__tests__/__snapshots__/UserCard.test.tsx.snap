// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard matches snapshot with admin user 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderBottomWidth": 1,
      "borderColor": "#E0E0E0",
      "marginBottom": 8,
      "marginHorizontal": 16,
      "paddingLeft": 0,
      "paddingVertical": 16,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#e8f2fb",
            "borderRadius": 14,
            "height": 50,
            "justifyContent": "center",
            "marginRight": 16,
            "width": 50,
          }
        }
      >
        <Text
          style={
            {
              "color": "#0171ce",
              "fontSize": 16,
              "fontWeight": "700",
              "marginBottom": 4,
            }
          }
        >
          JD
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#2C3E50",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          John Doe
        </Text>
        <Text
          style={
            {
              "color": "#7F8C8D",
              "fontSize": 14,
              "fontWeight": "600",
            }
          }
        >
          Admin
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#e8f2fb",
            "borderColor": "#0171ce",
            "borderRadius": 18,
            "borderWidth": 1,
            "height": 30,
            "justifyContent": "center",
            "opacity": 1,
            "width": 30,
          }
        }
        testID="usercard-edit"
      >
        <Icon
          color="#0171ce"
          name="edit"
          size={18}
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FDEDED",
            "borderColor": "#E74C3C",
            "borderRadius": 18,
            "borderWidth": 1,
            "height": 30,
            "justifyContent": "center",
            "opacity": 1,
            "width": 30,
          }
        }
        testID="usercard-delete"
      >
        <Icon
          color="#E74C3C"
          name="delete"
          size={18}
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`UserCard matches snapshot with manager user 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderBottomWidth": 1,
      "borderColor": "#E0E0E0",
      "marginBottom": 8,
      "marginHorizontal": 16,
      "paddingLeft": 0,
      "paddingVertical": 16,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#e8f2fb",
            "borderRadius": 14,
            "height": 50,
            "justifyContent": "center",
            "marginRight": 16,
            "width": 50,
          }
        }
      >
        <Text
          style={
            {
              "color": "#0171ce",
              "fontSize": 16,
              "fontWeight": "700",
              "marginBottom": 4,
            }
          }
        >
          JD
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            {
              "color": "#2C3E50",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          John Doe
        </Text>
        <Text
          style={
            {
              "color": "#7F8C8D",
              "fontSize": 14,
              "fontWeight": "600",
            }
          }
        >
          Manager
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 10,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#e8f2fb",
            "borderColor": "#0171ce",
            "borderRadius": 18,
            "borderWidth": 1,
            "height": 30,
            "justifyContent": "center",
            "opacity": 1,
            "width": 30,
          }
        }
        testID="usercard-edit"
      >
        <Icon
          color="#0171ce"
          name="edit"
          size={18}
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FDEDED",
            "borderColor": "#E74C3C",
            "borderRadius": 18,
            "borderWidth": 1,
            "height": 30,
            "justifyContent": "center",
            "opacity": 1,
            "width": 30,
          }
        }
        testID="usercard-delete"
      >
        <Icon
          color="#E74C3C"
          name="delete"
          size={18}
        />
      </View>
    </View>
  </View>
</View>
`;
